src/mockgyver.erl

```diff
@@ -1372,73 +1372,103 @@ ensure_renamed_mod_to_load_aux(RenamedMod, #modinfo{code=Code}) ->
+            mk_rewrite_st_function(Mod)]).
 
 mk_new_mod(Mod, ExportedFAs) ->
-    mk_mod(Mod, mk_new_funcs(Mod, ExportedFAs)).
-
-mk_new_funcs(Mod, ExportedFAs) ->
-    lists:map(fun(ExportedFA) -> mk_new_func(Mod, ExportedFA) end,
-              ExportedFAs).
-
-mk_new_func(Mod, {F, A}) ->
-    %% Generate a function like this (mod, func and arguments vary):
+    FmtNoAction =
+        fun(FnName, Args) ->
+                f("error_handler:raise_undef_exception(~p, ~s, ~s)",
+                  [Mod, FnName, Args])
+        end,
+    mk_mod(Mod,
+           [mk_handle_undefined_function(Mod, ExportedFAs, FmtNoAction),
+            mk_rewrite_st_function(Mod)]).
+
+mk_handle_undefined_function(Mod, ExportedFAs, FmtNoAction) ->
+    %% Parsing the string is approx 20% slower than constructing
+    %% the syntax tree using erl_syntax calls.
+    %% The string version is easier to understand though.
     %%
-    %%     func(A2, A1) ->
-    %%         case mockgyver:reg_call_and_get_action({mod,func,[A2, A1]}) of
-    %%             undefined ->
-    %%                 erlang:error(undef); % emulate undefined function
-    %%             ActionFun ->
-    %%                 ActionFun(A2, A1)
-    %%         end.
-    Args = mk_args(A),
-    Body =[erl_syntax:case_expr(
-             mk_call(mockgyver, reg_call_and_get_action,
-                     [erl_syntax:tuple([erl_syntax:abstract(Mod),
-                                        erl_syntax:abstract(F),
-                                        erl_syntax:list(Args)])]),
-             [erl_syntax:clause([erl_syntax:atom(undefined)],
-                                none,
-                                [mk_call(erlang, error,
-                                         [erl_syntax:atom(undef)])]),
-              erl_syntax:clause([erl_syntax:variable('ActionFun')],
-                                none,
-                                [mk_call('ActionFun', Args)])])],
-    erl_syntax:function(
-      erl_syntax:abstract(F),
-      [erl_syntax:clause(Args, none, Body)]).
+    %% It is many times faster than constructing a number of functions,
+    %% each containing the inner case expression, though.
+    func_from_str_fmt(
+      "'$handle_undefined_function'(FnName, Args) ->
+           Arity = length(Args),
+           case lists:member({FnName, Arity}, ~p) of
+               true ->
+                   case mockgyver:reg_call_and_get_action({~p,FnName,Args}) of
+                       undefined ->
+                           ~s;
+                       ActionFun ->
+                          try
+                              apply(ActionFun, Args)
+                          catch
+                              Class:Reason:St0 ->
+                                  ToMF = {~p, FnName},
+                                  St = rewrite_st(ActionFun, ToMF, Arity, St0),
+                                  erlang:raise(Class, Reason, St)
+                          end
```

Comment:
Nitpick: I noticed this block is indented by just 3

Reply:
Oups, indeed. Fixed in next revision.